contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -1687,3 +1695,112 @@ allocates resources such as `NodePorts` and `ClusterIPs` and automatically fill
+  This is safe because most users do not have acces to write to `status`.
+* PersistentVolumeClaims: Users may request a specific PersistentVolume in
+  `spec` or will be allocated one (in the same `spec` field).  If a specific PV
+  is requested, the volume controller will ensure that the volume is available
+  or else will report failure asynchronously.  Consumers read the result by
+  examining both the PVC and the PV.  This is more complicated than the others
+  because the spec value is stored before being confirmed, which could lead to
+  a user accessing someone else's PV.
+
+A counter-example:
+
+* Service `externalIPs`: Users must specify one or more specific IPs in `spec`.
+  The system cannot easily verify those IPs (by their definition, they are
+  external). Consumers read the result from `spec`.  This is UNSAFE and has
+  caused problems with untrusted users.
+
+In the past, API conventions dictated that `status` fields always come from
+observation, which made some of these cases more complicated than necessary.
+The conventions have been updated to allow `status` to hold such allocated
+values.  This is not a one-size-fits-all solution, though.
+
+### When to use a `spec` field
+
+New APIs should almost never do this.  Instead, they should use `status`.
+PersistentVolumes might have been simpler if we had done this.
+
+### When to use a `status` field
+
+Storing such values in `status` is the easiest and most straight-forward
+pattern.  This is appropriate when:
+
+* the allocated value is highly coupled to the rest of the object (e.g. pod
+  resource allocations)
+* the allocated value is always or almost always needed (i.e. most instances of
+  this type will have a value)
+* the schema and controller are known a priori (i.e. it's not an extension)
+* the access-control needs allow for controllers to be able to write to
+  `status`
+
+Consumers of such values can look at the `status` field for the "final" value.
+
+#### Sequencing operations
+
+Since almost everything is happening asynchronously to almost everything else,
+controller implementations should take care around the ordering of operations.
+For example, whether the controller updates a `status` field before or after it
+actuates a change depends on what guarantees need to be made to observers of
+the system.  In some cases, writing to a `status` field represents an
+acknowledgement or acceptance of a spec value, and it is OK to write it before
+actuation.  However, if it would be problematic for a client to observe the
+`status` value before it is actuated then the controller must actuate first and
+update `status` thereafter.  In some rarer cases, controllers will need to
+acknowledge, then actuate, then update to a "final" value.
+
+Controllers must take care to consider how a `status` field will be handled in
+the case of interrupted control loops (e.g. controller crash and restart), and
+must act idempotently and consistently.
```

Comment:
Do we want to recommend that controllers that write `status` before actuation should have some form of Condition that indicates that the change is not yet complete, and then clear that Condition in the final `status` update?

Reply:
Do we have a solid body of evidence to make specific prescriptions?  I don't know what the "right" recipe is in the abstract.  Maybe we can followup with more details?

Reply:
Most of my motivation here is trying to actuate these types of systems in automation (i.e. integration tests, installation).

In those situations, without a solid "it's done" signal, debugging failures becomes much harder because the orchestrator ends up needing to probe the resource to determine whether the actuation is complete. This basically ends up shredding the abstraction and introducing uncertainty between "true failures" and "didn't wait long enough" failures.

Reply:
Especially in layered systems, ESPECIALLY in networking, this is hard - it's not always  possible to know "it's done" without probing.   Again, I ama not sure what the prescription here should be given the need to be somewhat adaptable.